src/print_mem.c

```
@@ -0,0 +1,221 @@
+// vim:ts=4:sw=4:expandtab
+#include <ctype.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <yajl/yajl_gen.h>
+#include <yajl/yajl_version.h>
+#include "i3status.h"
+
+#define BINARY_BASE UINT64_C(1024)
+#define DECIMAL_BASE UINT64_C(1000)
+
+#define MAX_EXPONENT 4
+
+static const char filename[] = "/proc/meminfo";
+
+static const char *const iec_symbols[MAX_EXPONENT + 1] = {"", "Ki", "Mi", "Gi", "Ti"};
+static const char *const si_symbols[MAX_EXPONENT + 1] = {"", "k", "M", "G", "T"};
+static const char *const custom_symbols[MAX_EXPONENT + 1] = {"", "K", "M", "G", "T"};
+
+long get_long(const char *str) {
+    /* First skip non-digit characters */
+    /* Special case to handle negative numbers and the `+` sign */
+    while (*str && !(isdigit(*str) || ((*str == '-' || *str == '+') && isdigit(*(str + 1)))))
+        str++;
+    return strtol(str, NULL, 10);
+}
+
+/*
+ * Formats bytes according to the given base and set of symbols.
+ *
+ */
+static int format_bytes(char *outwalk, uint64_t bytes, uint64_t base, const char *const symbols[]) {
+    double size = bytes;
+    int exponent = 0;
+    while (size >= base && exponent < MAX_EXPONENT) {
+        size /= base;
+        exponent += 1;
+    }
+    return sprintf(outwalk, "%.1f %sB", size, symbols[exponent]);
+}
+
+/*
+ * Prints the given amount of bytes in a human readable manner.
+ *
+ */
+static int print_bytes_human(char *outwalk, uint64_t bytes, const char *prefix_type) {
+    if (strcasecmp(prefix_type, "decimal") == 0) {
+        return format_bytes(outwalk, bytes, DECIMAL_BASE, si_symbols);
+    } else if (strcasecmp(prefix_type, "custom") == 0) {
+        return format_bytes(outwalk, bytes, BINARY_BASE, custom_symbols);
+    } else {
```

Please use specific cases only and error out on invalid configurations.